File: src/main/scala/demos/fp/order/smartconstructor/Order.scala
package demos.fp.order.smartconstructor

class Order private(val lines: List[OrderLine] = List(), val orderId: String):
    def add(productId: String, amount: Int): Order =
        val existingOrderLineAndIndex = this.lines.zipWithIndex.find(_._1.productId == productId)
        val updatedLines = existingOrderLineAndIndex match
            case None =>
                val newLine = OrderLine(productId, amount)
                this.lines.appended(newLine)
            case Some((existingOrderLine, index)) =>
                val updatedLine = existingOrderLine.add(amount)
                this.lines.updated(index, updatedLine)
        new Order(lines = updatedLines, orderId=this.orderId)


object Order:
    def apply(orderId: String): Order = new Order(orderId = orderId)

File: src/main/java/demos/oop/order/smartconstructor/Order.java
package demos.oop.order.smartconstructor;

import java.util.ArrayList;
import java.util.List;

class Order {
    private List<OrderLine> lines;
    private String orderId;

    public Order(String orderId) {
        this.orderId = orderId;
        this.lines = new ArrayList<OrderLine>();
    }

    public void add(String productId, int amount) {
        OrderLine existingOrderLine = this.findExistingOrderLine(productId);
        if (existingOrderLine == null) {
            this.lines.add(new OrderLine(productId, amount));
        } else {
            existingOrderLine.add(amount);
        }
    }

    private OrderLine findExistingOrderLine(String productId) {
        for (OrderLine orderLine : this.lines) {
            if (orderLine.getProductId().equals(productId)) {
                return orderLine;
            }
        }
        return null;
    }

    public List<OrderLine> getLines() {
        return List.copyOf(this.lines);
    }
    
    public String getOrderId() {
        return orderId;
    }
}
